### PR TITLE
Add snarky generic add gadget

### PR DESCRIPTION
**Pull request by jspada**
This PR adds a snarky gadget for creating add gates with the generic gate.  It includes a new `Gadgets` module and adds some test cases for the new gadget.

### CODE REVIEW

**Comment by jspada** [comment]
!ci-build-me

File: src/lib/crypto/kimchi_backend/gadgets/generic.ml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      true
+    with _ -> false
+  in
+
+  (* Positive tests *)
+  assert (Bool.equal (test_generic_add 0 0 0) true) ;
+  assert (Bool.equal (test_generic_add 1 2 3) true) ;
+  (* Negatve tests *)
+  assert (Bool.equal (test_generic_add 1 0 0) false) ;
+  assert (Bool.equal (test_generic_add 2 4 7) false) ;
```
Comment (rbonichon) [inline src/lib/crypto/kimchi_backend/gadgets/generic.ml:77]: I would tend to have 4 different `let%test_unit` for these 4 tests, for example within a `let%test_module`. Do you think it is worthwhile to do so?

**Review by rbonichon — APPROVED** [review]

File: src/lib/crypto/kimchi_backend/gadgets/generic.ml (hunk 1 above)
Comment (rbonichon) [reply src/lib/crypto/kimchi_backend/gadgets/generic.ml:77]: See https://github.com/MinaProtocol/mina/blob/ccd92e1773344b069695b36d095135b2e1e271dd/src/lib/pickles/wrap.ml#L293

**Review by querolita — APPROVED** [review]

File: src/lib/crypto/kimchi_backend/gadgets/generic.ml (hunk 1 above)
Comment (jspada) [reply src/lib/crypto/kimchi_backend/gadgets/generic.ml:77]: I'm up for learning new things :-)

**Comment by mimoo** [comment]
what is the purpose of this gadget here : o ? we already have `Impl.Field.add` for adding two field elements together

**Comment by querolita** [comment]
Answering @mimoo, this was just for the sake of learning

File: src/lib/crypto/kimchi_backend/gadgets/generic.ml (hunk 1 above)
Comment (jspada) [reply src/lib/crypto/kimchi_backend/gadgets/generic.ml:77]: If we adopt this approach, does it need to be invoked with the in-line test runner?

```
   dune exec src/lib/pickles/.pickles.inline-tests/inline_test_runner_pickles.exe \
   --profile=dev --display short -- inline-test-runner pickles -only-test wrap.ml
```
?

Comment (jspada) [reply src/lib/crypto/kimchi_backend/gadgets/generic.ml:77]: if so, that could be a problem.  Perhaps this is best for a subsequent PR?

Comment (rbonichon) [reply src/lib/crypto/kimchi_backend/gadgets/generic.ml:77]: Not necessarily. It works with `runtest` as well.

Comment (rbonichon) [reply src/lib/crypto/kimchi_backend/gadgets/generic.ml:77]: Create https://github.com/MinaProtocol/mina/issues/12751

**Comment by jspada** [comment]
!ci-build-me

**Comment by jspada** [comment]
!ci-build-me

**Comment by rbonichon** [comment]
I just pushed the addition of `ppx_version` to dune - there's a lint check that was failing because of that.

**Comment by rbonichon** [comment]
!ci-build-me

**Comment by jspada** [comment]
!ci-build-me

**Pull request merged**